python/lsst/rubintv/production/aos.py

```diff
@@ -0,0 +1,219 @@
+    ):
+        self.butler = butler
+        self.locationConfig = locationConfig
+        self.inputCollection = inputCollection
+        self.outputCollection = outputCollection
+        self.pipelineFile = pipelineFile
+        self.queueName = queueName
+        self.allowMissingDependencies = allowMissingDependencies
+
+        self.instrument = "LSSTComCamSim"
+        self.repo = locationConfig.comCamButlerPath.replace("/butler.yaml", "")
+        self.log = logging.getLogger("lsst.rubintv.production.DonutLauncher")
+        self.redisHelper = RedisHelper(butler=butler, locationConfig=locationConfig)
+        self.checkSetup()
+        self.numCoresToUse = 9
+
+        self.runningProcesses = []
+        self.lock = threading.Lock()
+
+    def checkSetup(self):
+        try:
+            import batoid  # noqa: F401
+            import danish  # noqa: F401
+
+            import lsst.donut.viz as donutViz  # noqa: F401
+            import lsst.ts.wep as tsWep  # noqa: F401
+        except ImportError:
+            if self.allowMissingDependencies:
+                pass
+            else:
+                raise RuntimeError("Missing dependencies - can't launch donut pipelines like this")
+
+    def _extractExposureIds(self, exposureBytes):
+        """Extract the exposure IDs from the byte string.
+
+        Parameters
+        ----------
+        exposureBytes : `bytes`
+            The byte string containing the exposure IDs.
+
+        Returns
+        -------
+        expIds : `list` of `int`
+            A list of two exposure IDs extracted from the byte string.
+
+        Raises
+        ------
+        ValueError
+            If the number of exposure IDs extracted is not equal to 2.
+        """
+        exposureIds = exposureBytes.decode("utf-8").split(",")
+        exposureIds = [int(v) for v in exposureIds]
+        if len(exposureIds) != 2:
+            raise ValueError(f"Expected two exposureIds, got {exposureIds}")
+        return exposureIds
+
+    def _run_command(self, command):
+        """Run a command as a subprocess.
+
+        Runs the specified command as a subprocess, storing the process on the
+        class in a thread-safe manner. It logs the start time, process ID
+        (PID), and waits for the command to complete. If the command fails
+        (return code is non-zero), it logs an error. Finally, it logs the
+        completion time and duration of the command execution.
+
+        Parameters
+        ----------
+        command : `str`
+            The command to be executed.
+        """
+        start_time = time()
+        process = subprocess.Popen(command, shell=False)
+        with self.lock:
+            self.runningProcesses.append(process)
+        self.log.info(f"Process started with PID {process.pid}")
+        retcode = process.wait()
+        end_time = time()
+        duration = end_time - start_time
+        with self.lock:
+            self.runningProcesses.remove(process)
```

Comment:
Interesting, I never thought of that. It's a class instance, and seems pretty full featured (it might even just compare on PID in the background) but I've changed to a dict, thanks.